### ISSUE TITLE
C2C3 disc detection is failing in some data

### ISSUE THREAD

**Issue by jcohenadad** [spinalcordtoolbox/spinalcordtoolbox#2237]
In some cases, automatic c2c3 detection fails. The purpose of this issue is to list the problematic data, in order to improve the model/methods:
- sct_testing/bids/sub-20190418

To reproduce:
~~~
cd sub-20190418/anat/
sct_deepseg_sc -i sub-20190418_T2w.nii.gz -c t2
sct_label_vertebrae -i sub-20190418_T2w.nii.gz -s sub-20190418_T2w_seg.nii.gz -c t2 -r 0 -v 2
~~~

**Comment by charleygros** [comment]
I had a look at subject `sct_testing/bids/sct-users/sub-20190418`.

Wondering if the detection fails because of the detection itself or because of the post-processing.
Running:
```
fsleyes data_midSlice.nii pred_midSlice_before_postPro.nii.gz -cm red
```
We can see that, in that case, the detection itself (ie output of the model) fails. The post-processing can't really help much this case.

Note: the MR contrast of this subject seems quite different to the ones used in the training. If we want to improve our models, it could be relevant to include this subject into the training set.

**Comment by jcohenadad** [comment]
@charleygros running `sct_label_vertebrae` with `-r 0`, I don't see any of those files: `data_midSlice.nii`, `pred_midSlice_before_postPro.nii.gz`, because (i) `verbose` param is not passed on the `detect_c2c3()` call, and (ii) it should ideally be dealt with the param `remove_temp_folder`. I can fix that quickly

**Comment by charleygros** [comment]
@jcohenadad : in this PR, the verbose param is passed on the detect_c2c3 call.
Hope it helps

**Comment by jcohenadad** [comment]
> this PR

which one? I don't see any PR associated to this issue in https://github.com/neuropoly/spinalcordtoolbox/pulls. Or maybe i misunderstood?

**Comment by charleygros** [comment]
(oupsi sorry..)

This one: https://github.com/neuropoly/spinalcordtoolbox/pull/2296

**Comment by jcohenadad** [comment]
@charleygros looking at [your comment](https://github.com/neuropoly/spinalcordtoolbox/issues/2237#issuecomment-499046058), the files and the code, I am a bit confused as to what is called a "prediction", and what exactly the post-processing does. Is the file `data_midSlice_pred_svm` what you call the prediction? if so, how can you tell that the prediction fails in that case? (what are we supposed to expect from this output?). Maybe adding some documentation in the `detect_c2c3` file would be welcome.

**Comment by charleygros** [comment]
@jcohenadad I agree with you, will add some documentation in the `detect_c2c3`.
- Detection --> output of the SVM, ie `pred_midSlice_before_postPro.nii.gz` (which is the nifti associated with `data_midSlice_pred_svm`).
- How can we tell that the SVM failed? By visualising the following:
```
fsleyes data_midSlice.nii pred_midSlice_before_postPro.nii.gz -cm red -dr 0 100
```
... if there is no "red cluster" in the neighbourhood of C2-C3, then the detection failed.
... if there is a "red cluster" in the neighbourhood of C2-C3, then the post-processing got a rid of a TP svm-detected cluster.

**Comment by jcohenadad** [comment]
ok got it! Thanks. I was wondering if it was a "hard" fail that an algorithm could assess (e.g. no cluster at all), because in this case, there is a cluster but it is badly located (i.e. only human eyes can assess that it "failed", not an algorithm (chicken-and-egg problem)).

**Comment by jcohenadad** [comment]
Will be (partly) fixed by #2679

**Comment by joshuacwnewton** [comment]
I think we can close this issue, as it is essentially a duplicate of https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3793 -- If we're solving one issue, we're going to be solving the other.

I've edited the description of https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3793 to more clearly link back to this issue, so that none of the information in this issue is lost.
